astropy/io/fits/_utils.pyx

```diff
@@ -0,0 +1,55 @@
+# cython: language_level=3
+from collections import OrderedDict
+
+cdef Py_ssize_t BLOCK_SIZE = 2880  # the FITS block size
+cdef Py_ssize_t CARD_LENGTH = 80
+cdef str VALUE_INDICATOR = '= '  # The standard FITS value indicator
+cdef str END_CARD = 'END' + ' ' * 77
+
+
+def parse_header(fileobj):
+    """The main method allowing to parse quickly a FITS header."""
+
+    cards = OrderedDict()
+    cdef list read_blocks = []
+    cdef int found_end = 0
+    cdef bytes block
+    cdef str header_str, block_str, card_image, keyword
+    cdef Py_ssize_t idx, end_idx, sep_idx
+
+    while found_end == 0:
+        # iterate on blocks
+        block = fileobj.read(BLOCK_SIZE)
+        if not block or len(block) < BLOCK_SIZE:
+            # header looks incorrect, raising exception to fall back to
+            # the full Header parsing
+            raise Exception
```

Comment:
Seems odd to raise a plain Exception. That way we probably swallow any exceptions that we don't expect when catching it.

I don't have a good idea though what makes sense here. Maybe `ValueError` or some custom exception like `FastHeaderParsingFailedException`? 😄

Reply:
In any case there is a fallback to the Header parsing. I'm fine with using a custom exception but this will not make a difference, we need to catch any exception in the fast reader and then call Header as it have a better reporting of errors.